pubsub/gcp/pubsub/pubsub.go

```diff
@@ -235,20 +266,76 @@ func (g *GCPPubSub) Subscribe(req pubsub.SubscribeRequest, handler pubsub.Handle
 }
 
 func (g *GCPPubSub) handleSubscriptionMessages(topic *gcppubsub.Topic, sub *gcppubsub.Subscription, handler pubsub.Handler) error {
-	err := sub.Receive(context.Background(), func(ctx context.Context, m *gcppubsub.Message) {
-		msg := &pubsub.NewMessage{
-			Data:  m.Data,
-			Topic: topic.ID(),
-		}
+	// Limit the number of attempted reconnects we make.
+	reconnAttempts := make(chan struct{}, g.metadata.MaxReconnectionAttempts)
+	for i := 0; i < g.metadata.MaxReconnectionAttempts; i++ {
+		reconnAttempts <- struct{}{}
+	}
 
-		err := handler(ctx, msg)
+	readReconnectAttemptsRemaining := func() int { return len(reconnAttempts) }
+
+	// Periodically refill the reconnect attempts channel to avoid
+	// exhausting all the refill attempts due to intermittent issues
+	// occurring over a longer period of time.
+	reconnCtx, reconnCancel := context.WithCancel(g.ctx)
+	defer reconnCancel()
+
+	go func() {
+		// Calculate refill interval relative to MaxReconnectionAttempts and ConnectionRecoveryInSec
+		// in order to allow reconnection attempts to be exhausted in case of a permanent issue (e.g. wrong subscription name).
+		refillInterval := time.Duration(2*g.metadata.MaxReconnectionAttempts*g.metadata.ConnectionRecoveryInSec) * time.Second
+		for {
+			select {
+			case <-reconnCtx.Done():
+				g.logger.Debugf("Reconnect context for subscription %s is done", sub.ID())
+				return
+			case <-time.After(refillInterval):
+				attempts := readReconnectAttemptsRemaining()
+				if attempts < g.metadata.MaxReconnectionAttempts {
+					reconnAttempts <- struct{}{}
+				}
+				g.logger.Debugf("Number of reconnect attempts remaining for subscription %s: %d", sub.ID(), attempts)
+			}
+		}
+	}()
+
+	// Reconnect loop.
+	var receiveErr error = nil
+	for {
+		receiveErr = sub.Receive(g.ctx, func(ctx context.Context, m *gcppubsub.Message) {
```

Comment:
For my understanding, is the Subscription object reusable on err from `Receive` (i.e. should the Subscription object be recreated for the retry?)

Reply:
good question - yes, I have tested that messages are still received after the subscription is reconnected (with the code as is, where it's reusing the Subscription object instead of creating a new one)

Reply:
I have also added an issue to update the docs and linked in the description of this PR above